### PR TITLE
Correct llava mask & fix missing setter for `vocab_size`

### LINKED ISSUE #28184: LLaVa Left Padding Got Weird Results
### System Info

Reproduce : 

```python
from PIL import Image
import requests
from transformers import AutoProcessor, LlavaForConditionalGeneration

model = LlavaForConditionalGeneration.from_pretrained("llava-hf/llava-1.5-7b-hf").to(
    "cuda"
)
processor = AutoProcessor.from_pretrained("llava-hf/llava-1.5-7b-hf")

prompt1 = "<image>\n<image>\nUSER: What's the the difference of two images?\nASSISTANT:"
prompt2 = "<image>\nUSER: Describe the image.\nASSISTANT:"
prompt3 = "<image>\nUSER: Describe the image.\nASSISTANT:"
url1 = "https://images.unsplash.com/photo-1552053831-71594a27632d?q=80&w=3062&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D"
url2 = "https://images.unsplash.com/photo-1617258683320-61900b281ced?q=80&w=3087&auto=format&fit=crop&ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8fA%3D%3D"
image1 = Image.open(requests.get(url1, stream=True).raw)
image2 = Image.open(requests.get(url2, stream=True).raw)

inputs = processor(
    text=[prompt1, prompt2, prompt3],
    images=[image1, image2, image1, image2],
    return_tensors="pt",
    padding=True,
)
for key in inputs:
    inputs[key] = inputs[key].to("cuda")
    print(key, inputs[key].shape)

# Generate
generate_ids = model.generate(**inputs, max_length=512)
outputs = processor.batch_decode(
    generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False
)

print(outputs)
```

This will outputs : 
```Result
["\n \nUSER: What's the the difference of two images?\nASSISTANT: In the two images, the primary difference is the presence of a flower in the dog's mouth. In the first image, the dog is holding a flower in its mouth, while in the second image, the dog is not holding a flower. This subtle change in the scene highlights the dog's interaction with the flower, and it may evoke different emotions or interpretations depending on the viewer's perspective.", '\nUSER: Describe the image.\nASSISTANT: The dog is a \n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n', '\nUSER: Describe the image.\nASSISTANT: The \n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\n\nЪ schließ']
```

I checked images are rightly placed. but for batch2 and 3 
It's consist of lots of padding (False x 583)
[False x 583, False, True x 576  , False, False, False, False, False, False, False, False, False, False, False, False, False, False]

I guess llava doesn't see this kind of prefix on training phase would result in weird behavior.

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

stated at above

### Expected behavior

skip

### PR DESCRIPTION
**Pull request by fxmarty**
This PR fixes llava mask in the generation case.

`torch.cat((attention_mask, extended_mask), dim=-1)` was very wrong and it is a miracle we were getting meaningful generation for batch generation. The issue is that this disregards the attention_mask used during the first forward pass, that is custom. Wrong past key values are used.

It is also not clear to me why this custom masking is handled in the forward and not in GenerationMixin cc @gante (we are doing unnecessary operations at each forward to retrieve the correct attention_mask, it could just be an output of the model & updated in GenerationMixin)

Fixes https://github.com/huggingface/transformers/issues/28184

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Thanks! Could you add a test / update test values? Or has this no impact on the current tests?

**Comment by gante** [comment]
> It is also not clear to me why this custom masking is handled in the forward and not in GenerationMixin cc @gante (we are doing unnecessary operations at each forward to retrieve the correct attention_mask, it could just be an output of the model & updated in GenerationMixin)

@fxmarty We do that by default (in [this function](https://github.com/huggingface/transformers/blob/9c5e560924c4868033cc9ac650bee80dc4049158/src/transformers/generation/utils.py#L647)) 🤔 

@zucchini-nlp since you're interested in multimodal models: after this PR gets merged, can you inspect why llava (and related models) need this custom attention mask handling at generation time?

**Comment by zucchini-nlp** [comment]
@gante , I've been affected by this also while trying to make speculative decoding work with VLMs. I'll briefly outline what I found.

The custom mask is needed in Llava because it concatenated image embeds and text, which is further used as `inputs_embeds` for the LLM part. Thus after pre-fill they require attention mask of much larger length than what we got initially. When we update `model_kwargs` in `generate`, we assume one token is added to prev attention mask. This is not the case for Llava, since the attention mask holds values for textual part only, so the `forward` adds masks for the image part every call.

Comparing to other models, I have counted only 3 soft prompt VLMs in `transformers` (Kosmos-2, Blip, Llava). The other two models do not have this issue, because they have custom `generate` where embeds are concatenated, new attention masks are prepared manually and only then `text_model.generate()` is called for further handling. But Llava calls `composite_model.generate()` directly.

We can either fix Llava to call `text_model.generate()` for consistency with others or update `model_kwargs` depending on `past_key_values` length in `generate`. I was also thinking that if the second case, we can get rid of using `_prepare_token_type_ids` and `_prepare_attention_mask` in speculative decoding. 

It's up to you to decide if we need any changes, I have no idea how this can affect library-wide 😄

File: tests/models/llava/test_modeling_llava.py [hunk 1]
```diff
@@ -307,10 +307,50 @@ def test_small_model_integration_test_llama_batched_regression(self):
 
         output = model.generate(**inputs, max_new_tokens=20)
 
-        EXPECTED_DECODED_TEXT = ['USER:  \nWhat are the things I should be cautious about when I visit this place? What should I bring with me?\nASSISTANT: When visiting this serene location, one should be cautious about the weather conditions and potential', 'USER:  \nWhat is this?\nASSISTANT: Two cats lying on a bed!\nUSER:  \nAnd this?\nASSISTANT: A cat sleeping on a bed.']  # fmt: skip
+        EXPECTED_DECODED_TEXT = ['USER:  \nWhat are the things I should be cautious about when I visit this place? What should I bring with me?\nASSISTANT: When visiting this place, which appears to be a dock or pier extending over a body of water', 'USER:  \nWhat is this?\nASSISTANT: Two cats lying on a bed!\nUSER:  \nAnd this?\nASSISTANT: A cat sleeping on a bed.']  # fmt: skip
```
Comment (fxmarty) [inline tests/models/llava/test_modeling_llava.py:310]: This test itself was wrong. The input_ids is `tensor([[32001, 32001, 32001,     1,  3148,  1001, 29901, 29871, 32000, 29871,
            13,  5618,   526,   278,  2712,   306,   881,   367,   274,  1300,
          2738,  1048,   746,   306,  6493,   445,  2058, 29973,  1724,   881,
           306,  6963,   411,   592, 29973,    13, 22933,  9047, 13566, 29901],
        [    1,  3148,  1001, 29901, 29871, 32000, 29871,    13,  5618,   338,
           445, 29973,    13, 22933,  9047, 13566, 29901,  7803,   274,  1446,
         19214,   373,   263,  6592, 29991,    13, 11889, 29901, 29871, 32000,
         29871,    13,  2855,   445, 29973,    13, 22933,  9047, 13566, 29901]],
       device='cuda:0')`, with `32000` being the image tokens. The first sequences visibly see only the first image, so the output should not be different compared to https://github.com/huggingface/transformers/blob/00c1d87a7d5c8dfb4554370983b5a3f7c069edd7/tests/models/llava/test_modeling_llava.py#L245-L265, should it @younesbelkada?

File: tests/models/llava/test_modeling_llava.py [hunk 2]
```diff
@@ -307,10 +307,50 @@ def test_small_model_integration_test_llama_batched_regression(self):
...
 
         self.assertEqual(processor.batch_decode(output, skip_special_tokens=True), EXPECTED_DECODED_TEXT)
 
+    @slow
+    def test_batched_generation(self):
```
Comment (fxmarty) [inline tests/models/llava/test_modeling_llava.py:315]: This test does not pass on main (trash is generated instead)

File: src/transformers/models/llava/modeling_llava.py [hunk 3]
```diff
@@ -340,6 +340,12 @@ def _merge_input_ids_with_image_features(self, image_features, inputs_embeds, in
         final_attention_mask |= image_to_overwrite
         position_ids = (final_attention_mask.cumsum(-1) - 1).masked_fill_((final_attention_mask == 0), 1)
 
+        # 6. Mask out the embedding at padding positions, as we later use the past_key_value value to determine the non-attended tokens.
+        batch_indices, pad_indices = torch.where(input_ids == self.pad_token_id)
+        indices_to_mask = new_token_positions[batch_indices, pad_indices]
+
+        final_embedding[batch_indices, indices_to_mask] = 0
+
```
Comment (fxmarty) [inline src/transformers/models/llava/modeling_llava.py:348]: As later `batch_index, non_attended_tokens = torch.where(first_layer_past_key_value.float().sum(-2) == 0)` is used, this is necessary

**Comment by fxmarty** [comment]
Llava slow tests all pass (apart from `tests/models/llava/test_modeling_llava.py::LlavaForConditionalGenerationModelTest::test_cpu_offload` which does not pass on main either).

**Review by gante — APPROVED** [review]
The changes look good to me 👍

**Review by younesbelkada — APPROVED** [review]
Thanks so much for the fix and the deep investigation ! LGTM since the slow tests pass

File: src/transformers/models/llava/modeling_llava.py [hunk 4]
```diff
@@ -340,6 +340,12 @@ def _merge_input_ids_with_image_features(self, image_features, inputs_embeds, in
         final_attention_mask |= image_to_overwrite
         position_ids = (final_attention_mask.cumsum(-1) - 1).masked_fill_((final_attention_mask == 0), 1)
 
+        # 6. Mask out the embedding at padding positions, as we later use the past_key_value value to determine the non-attended tokens.
+        batch_indices, pad_indices = torch.where(input_ids == self.pad_token_id)
+        indices_to_mask = new_token_positions[batch_indices, pad_indices]
+
+        final_embedding[batch_indices, indices_to_mask] = 0
```
Comment (ArthurZucker) [inline src/transformers/models/llava/modeling_llava.py:347]:
```suggestion
        indices_to_mask = new_token_positions * (input_ids != self.pad_token_id)
```
should this not work as well?

**Review by ArthurZucker — COMMENTED** [review]
LGTM, might be a simpler way to fix this

File: src/transformers/models/llava/modeling_llava.py (hunk 4 above)
Comment (fxmarty) [reply src/transformers/models/llava/modeling_llava.py:347]: maybe

**Comment by fxmarty** [comment]
I added a setter to fix the regression in https://github.com/huggingface/transformers/pull/29586#discussion_r1535058927

File: src/transformers/models/llava/configuration_llava.py [hunk 5]
```diff
@@ -147,6 +147,10 @@ def vocab_size(self):
         )
         return self._vocab_size
 
+    @vocab_size.setter
+    def vocab_size(self, value):
+        self._vocab_size = value
+
```
Comment (fxmarty) [inline src/transformers/models/llava/configuration_llava.py:153]: Is this good @NielsRogge @amyeroberts ?

**Review by amyeroberts — APPROVED** [review]
Thanks for digging into this and fixing! 

Overall the changes look OK, I'm just a bit unclear on some of the reasoning for the logic changes

File: src/transformers/models/llava/configuration_llava.py (hunk 5 above)
Comment (NielsRogge) [reply src/transformers/models/llava/configuration_llava.py:153]: Shouldn't we remove the `@property` annotator above? That fixed https://github.com/huggingface/transformers/issues/29789 for me. Isn't property for immutable things (whereas the vocabulary size can change? cc @amyeroberts

Comment (amyeroberts) [reply src/transformers/models/llava/configuration_llava.py:153]: Properties aren't for immutable things (this is the value of a setter). The property enables us to emit a deprecation warning if the property is accessed as an attribute.

Removing the property annotation would mean that previous usage of `config.vocab_size` would break, and people would have to use `config.vocab_size()`

File: tests/models/llava/test_modeling_llava.py [hunk 6]
```diff
@@ -307,10 +307,50 @@ def test_small_model_integration_test_llama_batched_regression(self):
...
+        import requests
+        import torch
+        from PIL import Image
+
+        from transformers import AutoProcessor, LlavaForConditionalGeneration
+
+        with torch.device(torch_device):
+            model = LlavaForConditionalGeneration.from_pretrained("llava-hf/llava-1.5-7b-hf")
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply tests/models/llava/test_modeling_llava.py:486]: It is faster with the context manager.

File: src/transformers/models/llava/modeling_llava.py [hunk 7]
```diff
@@ -344,6 +344,12 @@ def _merge_input_ids_with_image_features(self, image_features, inputs_embeds, in
         final_attention_mask |= image_to_overwrite
         position_ids = (final_attention_mask.cumsum(-1) - 1).masked_fill_((final_attention_mask == 0), 1)
 
+        # 6. Mask out the embedding at padding positions, as we later use the past_key_value value to determine the non-attended tokens.
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply src/transformers/models/llava/modeling_llava.py:347]: A bit later in the code (specifically here: https://github.com/huggingface/transformers/blob/aa17cf986f0761382f3d6e591e985a42671c3fb7/src/transformers/models/llava/modeling_llava.py#L449), we check `past_key_values` value to find unattended tokens. This was not correct before because this added step `6.` was missing.

Overall, this would be worth a larger refactor that would avoid regenerating full masks at every forward step in the generate. This PR is just a hotfix.

**Comment by fxmarty** [comment]
@amyeroberts We need to set `final_embedding` to zero for https://github.com/huggingface/transformers/blob/aa17cf986f0761382f3d6e591e985a42671c3fb7/src/transformers/models/llava/modeling_llava.py#L449

to do its job!

File: tests/models/llava/test_modeling_llava.py [hunk 8]
```diff
@@ -307,10 +307,50 @@ def test_small_model_integration_test_llama_batched_regression(self):
...
 
         self.assertEqual(processor.batch_decode(output, skip_special_tokens=True), EXPECTED_DECODED_TEXT)
 
+    @slow
+    def test_batched_generation(self):
+        import requests
+        import torch
+        from PIL import Image
+
+        from transformers import AutoProcessor, LlavaForConditionalGeneration
```
[the remark this answers is not in the document]
Comment (fxmarty) [reply tests/models/llava/test_modeling_llava.py:483]: thank you!
